src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,141 @@
+
+        let report = tee_evidence.attestation_report;
+        if report.version != 2 {
+            return Err(anyhow!("Unexpected report version"));
+        }
+
+        if report.vmpl != 0 {
+            return Err(anyhow!("VMPL Check Failed"));
+        }
+
+        let expected_report_data = calculate_expected_report_data(&nonce, &attestation.tee_pubkey);
+        if report.report_data != expected_report_data {
+            return Err(anyhow!("Report Data Mismatch"));
+        }
+
+        parse_tee_evidence(&report)
+    }
+}
+
+fn verify_report_signature(evidence: &SnpEvidence) -> Result<()> {
+    // check cert chain
+    let vcek = verify_cert_chain(&evidence.cert_chain)?;
+
+    // verify report signature
+    let sig = ecdsa::EcdsaSig::try_from(&evidence.attestation_report.signature)?;
+    let data = &bincode::serialize(&evidence.attestation_report)?[..=0x29];
+
+    sig.verify(&data, EcKey::try_from(vcek.public_key()?)?.as_ref())
+        .map_err(|_| anyhow!("Signature validation failed."))?;
+
+    Ok(())
+}
+
+fn verify_cert_chain(cert_chain: &CertTable) -> Result<x509::X509> {
+    // add support for genoa
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ark_ask.pem"))?;
+    
+    // make this nicer
+    let ask = &certs[0];
+    let ark = &certs[1];
+
+    // check for revocation?
+
+    //let raw_vcek = cert_chain.entries.iter().find(|c| c.cert_type == SnpCertType::VCEK)
+    let raw_vcek = cert_chain
+        .entries
+        .iter()
+        .find(|c| c.guid() == "63da758d-e664-4564-adc5-f4b93be8accd")
+        .ok_or(anyhow!("VCEK not found."))?;
+    let vcek =
+        x509::X509::from_der(&raw_vcek.data()).map_err(|_| anyhow!("Failed to load VCEK"))?;
+
+    // ARK -> ASK
+    ask.verify(&(ark.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid ASK Signature"))?;
+
+    // ASK -> VCEK
+    vcek.verify(&(ask.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid VCEK Signature"))?;
+
+    Ok(vcek)
+}
+
+fn calculate_expected_report_data(nonce: &String, tee_pubkey: &String) -> [u8; 64] {
+    let mut hasher = Sha384::new();
+
+    hasher.update(nonce.clone().into_bytes());
+    hasher.update(tee_pubkey.clone().into_bytes());
+
+    let mut hash = hasher.finalize().to_vec();
+    hash.extend([0u8; 16]);
+
+    hash.as_slice().try_into().unwrap()
+}
+
+fn parse_tee_evidence(report: &AttestationReport) -> Result<TeeEvidenceParsedClaim> {
+    let claims_map = json!({
+        // policy fields
+        "policy_abi_major": report.policy.abi_major(),
+        "policy_abi_minor": report.policy.api_minor(),
+        "policy_smt_allowed": report.policy.smt_allowed(),
+        "policy_migrate_ma": report.policy.migrate_ma_allowed(),
+        "policy_debug_allowed": report.policy.debug_allowed(),
+        "policy_single_socket": report.policy.single_socket_required(),
+
+        // versioning info
+        "current_tcb_bootloader": report.current_tcb.boot_loader,
```

Comment:
I think the more important one to check is actually `reported_tcb`.

Reply:
I think we need to compare the VCEK to the reported TCB. I forgot to implement this. I am still making up my mind about which tcb version should actually be used as a claim.